TypeSystem/Serialization/test.cc

```diff
@@ -200,7 +216,129 @@ TEST(Serialization, Binary) {
 }
 #endif
 
-TEST(Serialization, JSON) {
+TEST(JSONSerialization, CPPTypes) {
+  // `bool`.
+  EXPECT_EQ("true", JSON(true));
+  EXPECT_TRUE(ParseJSON<bool>("true"));
+  ASSERT_THROW(ParseJSON<bool>("1"), JSONSchemaException);
+
+  EXPECT_EQ("false", JSON(false));
+  EXPECT_FALSE(ParseJSON<bool>("false"));
+  ASSERT_THROW(ParseJSON<bool>("0"), JSONSchemaException);
+  ASSERT_THROW(ParseJSON<bool>(""), InvalidJSONException);
+
+  // `int`.
+  EXPECT_EQ("42", JSON(42));
+  EXPECT_EQ(42, ParseJSON<int>("42"));
+
+  // `std::string`.
+  EXPECT_EQ("\"forty two\"", JSON("forty two"));
+  EXPECT_EQ("forty two", ParseJSON<std::string>("\"forty two\""));
+
+  EXPECT_EQ("\"a\\u0000b\"", JSON(std::string("a\0b", 3)));
+  EXPECT_EQ(std::string("c\0d", 3), ParseJSON<std::string>("\"c\\u0000d\""));
+
+  // `std::vector<>`.
+  EXPECT_EQ("[]", JSON(std::vector<uint64_t>()));
+  EXPECT_EQ("[1,2,3]", JSON(std::vector<uint64_t>({1, 2, 3})));
+  EXPECT_EQ("[[\"one\",\"two\"],[\"three\",\"four\"]]",
+            JSON(std::vector<std::vector<std::string>>({{"one", "two"}, {"three", "four"}})));
+  EXPECT_EQ(4u, ParseJSON<std::vector<std::vector<std::string>>>("[[],[],[],[]]").size());
+  EXPECT_EQ("blah", ParseJSON<std::vector<std::vector<std::string>>>("[[],[\"\",\"blah\"],[],[]]")[1][1]);
+
+  // `std::map<>`.
+  using map_int_int = std::map<int, int>;
+  using map_string_int = std::map<std::string, int>;
+  EXPECT_EQ("[]", JSON(map_int_int()));
+  EXPECT_EQ("{}", JSON(map_string_int()));
```

Comment:
Would be nice to add a comment here that string-as-key results in an object, while any other key type makes it an array of arrays of two elements.